From 4d774ad7b71342c5e3ba2f1f5ac9d051eb835270 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Wed, 9 Dec 2015 14:17:38 +0100
Subject: [PATCH 03/55] vnc: altgr emulation

---
 ui/vnc.c | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/ui/vnc.c b/ui/vnc.c
index 3e89dad..88a5824 100644
--- a/ui/vnc.c
+++ b/ui/vnc.c
@@ -1711,6 +1711,10 @@ static void kbd_leds(void *opaque, int ledstate)
 
 static void do_key_event(VncState *vs, int down, int keycode, int sym)
 {
+    int mods =  keycode & 0xf00;
+
+    keycode &= SCANCODE_KEYMASK;
+
     /* QEMU console switch */
     switch(keycode) {
     case 0x2a:                          /* Left Shift */
@@ -1791,7 +1795,24 @@ static void do_key_event(VncState *vs, int down, int keycode, int sym)
     }
 
     if (qemu_console_is_graphic(NULL)) {
+
+ 	/* our java vnc client never sends ALTGR, so we create
+	   an artificial up/down event */
+
+	int emul_altgr = (mods & SCANCODE_ALTGR) &&
+	    !vs->modifiers_state[0xb8];
+
+	if (emul_altgr) {
+            reset_keys(vs);
+            qemu_input_event_send_key_number(vs->vd->dcl.con, 0xb8, true);
+	}
+
         qemu_input_event_send_key_number(vs->vd->dcl.con, keycode, down);
+
+	if (emul_altgr) {
+             qemu_input_event_send_key_number(vs->vd->dcl.con, 0xb8, false);
+	}
+
     } else {
         bool numlock = vs->modifiers_state[0x45];
         bool control = (vs->modifiers_state[0x1d] ||
@@ -1930,7 +1951,8 @@ static void key_event(VncState *vs, int down, uint32_t sym)
         lsym = lsym - 'A' + 'a';
     }
 
-    keycode = keysym2scancode(vs->vd->kbd_layout, lsym & 0xFFFF) & SCANCODE_KEYMASK;
+    keycode = keysym2scancode(vs->vd->kbd_layout, lsym & 0xFFFF);
+
     trace_vnc_key_event_map(down, sym, keycode, code2name(keycode));
     do_key_event(vs, down, keycode, sym);
 }
-- 
2.1.4

