From 38b2b042f9943786c9d3728d9c24cb110170df88 Mon Sep 17 00:00:00 2001
From: Alexandre Derumier <aderumier@odiso.com>
Date: Tue, 29 Sep 2015 15:37:44 +0200
Subject: [PATCH 25/41] smm_available = false

Signed-off-by: Alexandre Derumier <aderumier@odiso.com>
---
 hw/i386/pc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/i386/pc.c b/hw/i386/pc.c
index ba8a5a1..9c206fc 100644
--- a/hw/i386/pc.c
+++ b/hw/i386/pc.c
@@ -2084,7 +2084,7 @@ bool pc_machine_is_smm_enabled(PCMachineState *pcms)
     if (tcg_enabled() || qtest_enabled()) {
         smm_available = true;
     } else if (kvm_enabled()) {
-        smm_available = kvm_has_smm();
+        smm_available = false;
     }
 
     if (smm_available) {
-- 
2.1.4

