From a826a6ead084ea0d86680c6e9901f4f51ee5beb0 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Tue, 12 Jan 2016 09:09:49 +0100
Subject: [PATCH 29/41] vnc: make x509 imply tls again

---
 ui/vnc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/ui/vnc.c b/ui/vnc.c
index 8d7d435..a7f5f1e 100644
--- a/ui/vnc.c
+++ b/ui/vnc.c
@@ -3725,9 +3725,8 @@ void vnc_display_open(const char *id, Error **errp)
         const char *path;
         bool tls = false, x509 = false, x509verify = false;
         tls  = qemu_opt_get_bool(opts, "tls", false);
-        if (tls) {
-            path = qemu_opt_get(opts, "x509");
-
+        path = qemu_opt_get(opts, "x509");
+        if (tls || path) {
             if (path) {
                 x509 = true;
             } else {
-- 
2.1.4

