From f51a30dbfc454e79b98c7a823db95d7b1532ea83 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Fri, 21 Oct 2016 09:09:26 +0200
Subject: [PATCH 43/47] vma: sizes passed to blk_co_preadv should be bytes now

---
 vma.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/vma.c b/vma.c
index 4f55799..0491542 100644
--- a/vma.c
+++ b/vma.c
@@ -466,8 +466,8 @@ static void coroutine_fn backup_run(void *opaque)
         iov.iov_len = VMA_CLUSTER_SIZE;
         qemu_iovec_init_external(&qiov, &iov, 1);
 
-        ret = blk_co_preadv(job->target, start * BACKUP_SECTORS_PER_CLUSTER,
-                            BACKUP_SECTORS_PER_CLUSTER, &qiov, 0);
+        ret = blk_co_preadv(job->target, start * VMA_CLUSTER_SIZE,
+                            VMA_CLUSTER_SIZE, &qiov, 0);
         if (ret < 0) {
             vma_writer_set_error(job->vmaw, "read error", -1);
             goto out;
-- 
2.1.4

