From 2415c407a55184041415874ff84f14b583d0e766 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Mon, 24 Oct 2016 09:32:36 +0200
Subject: [PATCH 44/47] glusterfs: no default logfile if daemonized

---
 block/gluster.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/block/gluster.c b/block/gluster.c
index a577dae..e712dc7 100644
--- a/block/gluster.c
+++ b/block/gluster.c
@@ -33,7 +33,7 @@
 #define GLUSTER_DEBUG_DEFAULT       4
 #define GLUSTER_DEBUG_MAX           9
 #define GLUSTER_OPT_LOGFILE         "logfile"
-#define GLUSTER_LOGFILE_DEFAULT     "-" /* handled in libgfapi as /dev/stderr */
+#define GLUSTER_LOGFILE_DEFAULT     NULL
 
 #define GERR_INDEX_HINT "hint: check in 'server' array index '%d'\n"
 
@@ -398,6 +398,7 @@ static struct glfs *qemu_gluster_glfs_init(BlockdevOptionsGluster *gconf,
     int old_errno;
     SocketAddressFlatList *server;
     unsigned long long port;
+    const char *logfile;
 
     glfs = glfs_find_preopened(gconf->volume);
     if (glfs) {
@@ -433,9 +434,15 @@ static struct glfs *qemu_gluster_glfs_init(BlockdevOptionsGluster *gconf,
         }
     }
 
-    ret = glfs_set_logging(glfs, gconf->logfile, gconf->debug);
-    if (ret < 0) {
-        goto out;
+    logfile = gconf->logfile;
+    if (!logfile && !is_daemonized()) {
+        logfile = "-";
+    }
+    if (logfile) {
+        ret = glfs_set_logging(glfs, logfile, gconf->debug);
+        if (ret < 0) {
+            goto out;
+        }
     }
 
     ret = glfs_init(glfs);
-- 
2.1.4

