From a7613eb93e702d5de5b40d17c4d4e95e8e5a010d Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Mon, 24 Oct 2016 09:32:36 +0200
Subject: [PATCH 44/47] glusterfs: daemonize

---
 block/gluster.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/block/gluster.c b/block/gluster.c
index 01b479f..6dcf926 100644
--- a/block/gluster.c
+++ b/block/gluster.c
@@ -341,9 +341,11 @@ static struct glfs *qemu_gluster_glfs_init(BlockdevOptionsGluster *gconf,
         }
     }
 
-    ret = glfs_set_logging(glfs, "-", gconf->debug_level);
-    if (ret < 0) {
-        goto out;
+    if (!is_daemonized()) {
+        ret = glfs_set_logging(glfs, "-", gconf->debug_level);
+        if (ret < 0) {
+            goto out;
+        }
     }
 
     ret = glfs_init(glfs);
-- 
2.1.4

