From 7878879d6dc2330e6d2d34c1e83101ad796ce939 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Tue, 12 Apr 2016 13:49:44 +0200
Subject: [PATCH] vma: also "guess" raw for /dev/ paths

---
 vma.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/vma.c b/vma.c
index 1c4103f..2f604bb 100644
--- a/vma.c
+++ b/vma.c
@@ -304,13 +304,15 @@ static int extract_content(int argc, char **argv)
 	    size_t devlen = strlen(devfn);
 	    bool protocol = path_has_protocol(devfn);
 	    QDict *options = NULL;
-	    if (devlen > 4 && strcmp(devfn+devlen-4, ".raw") == 0 && !protocol) {
+	    if (protocol) {
+		/* tell bdrv_open to honor the protocol */
+		flags |= BDRV_O_PROTOCOL;
+	    } else if ((devlen > 4 && strcmp(devfn+devlen-4, ".raw") == 0) ||
+	               strncmp(devfn, "/dev/", 5) == 0)
+	    {
 		/* explicit raw format */
 		options = qdict_new();
 		qdict_put(options, "driver", qstring_from_str("raw"));
-	    } else if (protocol) {
-		/* tell bdrv_open to honor the protocol */
-		flags |= BDRV_O_PROTOCOL;
 	    }
 
 	    if (errp || bdrv_open(&bs, devfn, NULL, options, flags, &errp)) {
-- 
2.1.4

