Index: new/net.h
===================================================================
--- new.orig/net.h	2012-09-24 07:15:00.000000000 +0200
+++ new/net.h	2012-09-24 07:15:16.000000000 +0200
@@ -161,8 +161,9 @@
 void netdev_add(QemuOpts *opts, Error **errp);
 int qmp_netdev_add(Monitor *mon, const QDict *qdict, QObject **ret);
 
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
+#define DEFAULT_NETWORK_SCRIPT "/etc/kvm/kvm-ifup"
+#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/kvm/kvm-ifdown"
+
 #define DEFAULT_BRIDGE_HELPER CONFIG_QEMU_HELPERDIR "/qemu-bridge-helper"
 #define DEFAULT_BRIDGE_INTERFACE "br0"
 
Index: new/configure
===================================================================
--- new.orig/configure	2012-09-24 07:15:00.000000000 +0200
+++ new/configure	2012-09-24 07:15:16.000000000 +0200
@@ -195,7 +195,7 @@
 libdir="\${prefix}/lib"
 includedir="\${prefix}/include"
 sysconfdir="\${prefix}/etc"
-confsuffix="/qemu"
+confsuffix="/kvm"
 slirp="yes"
 fmod_lib=""
 fmod_inc=""
Index: new/net/tap.h
===================================================================
--- new.orig/net/tap.h	2012-09-24 07:15:00.000000000 +0200
+++ new/net/tap.h	2012-09-24 07:15:16.000000000 +0200
@@ -29,8 +29,6 @@
 #include "qemu-common.h"
 #include "qapi-types.h"
 
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
 
 int net_init_tap(const NetClientOptions *opts, const char *name,
                  NetClientState *peer);
