Index: new/net.h
===================================================================
--- new.orig/net.h	2012-02-15 11:17:27.000000000 +0100
+++ new/net.h	2012-02-15 11:29:29.000000000 +0100
@@ -173,8 +173,9 @@
 int do_netdev_add(Monitor *mon, const QDict *qdict, QObject **ret_data);
 int do_netdev_del(Monitor *mon, const QDict *qdict, QObject **ret_data);
 
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
+#define DEFAULT_NETWORK_SCRIPT "/etc/kvm/kvm-ifup"
+#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/kvm/kvm-ifdown"
+
 #define DEFAULT_BRIDGE_HELPER CONFIG_QEMU_HELPERDIR "/qemu-bridge-helper"
 #define DEFAULT_BRIDGE_INTERFACE "br0"
 
Index: new/configure
===================================================================
--- new.orig/configure	2012-02-15 11:17:27.000000000 +0100
+++ new/configure	2012-02-15 11:21:15.000000000 +0100
@@ -167,7 +167,7 @@
 libdir="\${prefix}/lib"
 includedir="\${prefix}/include"
 sysconfdir="\${prefix}/etc"
-confsuffix="/qemu"
+confsuffix="/kvm"
 slirp="yes"
 fmod_lib=""
 fmod_inc=""
Index: new/net/tap.h
===================================================================
--- new.orig/net/tap.h	2012-02-15 11:17:27.000000000 +0100
+++ new/net/tap.h	2012-02-15 11:21:15.000000000 +0100
@@ -29,8 +29,6 @@
 #include "qemu-common.h"
 #include "qemu-option.h"
 
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
 
 int net_init_tap(QemuOpts *opts, Monitor *mon, const char *name, VLANState *vlan);
 
