Index: new/arch_init.c
===================================================================
--- new.orig/arch_init.c	2012-12-27 12:28:50.000000000 +0100
+++ new/arch_init.c	2012-12-27 12:29:24.000000000 +0100
@@ -662,6 +662,8 @@
     DPRINTF("ram_save_live: expected(%" PRIu64 ") <= max(" PRIu64 ")?\n",
             expected_downtime, migrate_max_downtime());
 
+    s->expected_downtime = expected_downtime / 1000000; /* ns -> ms */
+
     if (expected_downtime <= migrate_max_downtime()) {
         migration_bitmap_sync();
         expected_downtime = ram_save_remaining() * TARGET_PAGE_SIZE / bwidth;
