Index: new/block/backup.c
===================================================================
--- new.orig/block/backup.c	2014-07-16 11:09:01.000000000 +0200
+++ new/block/backup.c	2014-07-16 11:16:47.000000000 +0200
@@ -39,6 +39,7 @@
     BlockDriverState *target;
     MirrorSyncMode sync_mode;
     RateLimit limit;
+    BackupDumpFunc *dump_cb;
     BlockdevOnError on_source_error;
     BlockdevOnError on_target_error;
     CoRwlock flush_rwlock;
@@ -136,14 +137,21 @@
             goto out;
         }
 
+        int64_t start_sec = start * BACKUP_SECTORS_PER_CLUSTER;
         if (buffer_is_zero(iov.iov_base, iov.iov_len)) {
-            ret = bdrv_co_write_zeroes(job->target,
-                                       start * BACKUP_SECTORS_PER_CLUSTER,
-                                       n, BDRV_REQ_MAY_UNMAP);
+            if (job->dump_cb) {
+                ret = job->dump_cb(job->common.opaque, job->target, start_sec, n, NULL);
+            }
+            if (job->target) {
+                ret = bdrv_co_write_zeroes(job->target, start_sec, n, BDRV_REQ_MAY_UNMAP);
+            }
         } else {
-            ret = bdrv_co_writev(job->target,
-                                 start * BACKUP_SECTORS_PER_CLUSTER, n,
-                                 &bounce_qiov);
+            if (job->dump_cb) {
+                ret = job->dump_cb(job->common.opaque, job->target, start_sec, n, bounce_buffer);
+            }
+            if (job->target) {
+                ret = bdrv_co_writev(job->target, start_sec, n, &bounce_qiov);
+            }
         }
         if (ret < 0) {
             trace_backup_do_cow_write_fail(job, start, ret);
@@ -205,7 +213,9 @@
 {
     BackupBlockJob *s = container_of(job, BackupBlockJob, common);
 
-    bdrv_iostatus_reset(s->target);
+    if (s->target) {
+        bdrv_iostatus_reset(s->target);
+    }
 }
 
 static const BlockJobDriver backup_job_driver = {
@@ -221,9 +231,11 @@
     if (read) {
         return block_job_error_action(&job->common, job->common.bs,
                                       job->on_source_error, true, error);
-    } else {
+    } else if (job->target) {
         return block_job_error_action(&job->common, job->target,
                                       job->on_target_error, false, error);
+    } else {
+        return BLOCK_ERROR_ACTION_REPORT;
     }
 }
 
@@ -248,9 +260,11 @@
 
     job->bitmap = hbitmap_alloc(end, 0);
 
-    bdrv_set_enable_write_cache(target, true);
-    bdrv_set_on_error(target, on_target_error, on_target_error);
-    bdrv_iostatus_enable(target);
+    if (target) {
+        bdrv_set_enable_write_cache(target, true);
+        bdrv_set_on_error(target, on_target_error, on_target_error);
+        bdrv_iostatus_enable(target);
+    }
 
     bdrv_add_before_write_notifier(bs, &before_write);
 
@@ -343,8 +357,10 @@
 
     hbitmap_free(job->bitmap);
 
-    bdrv_iostatus_disable(target);
-    bdrv_unref(target);
+    if (target) {
+        bdrv_iostatus_disable(target);
+        bdrv_unref(target);
+    }
 
     block_job_completed(&job->common, ret);
 }
@@ -353,13 +369,15 @@
                   int64_t speed, MirrorSyncMode sync_mode,
                   BlockdevOnError on_source_error,
                   BlockdevOnError on_target_error,
+                  BackupDumpFunc *dump_cb,
                   BlockDriverCompletionFunc *cb, void *opaque,
+                  bool paused,
                   Error **errp)
 {
     int64_t len;
 
     assert(bs);
-    assert(target);
+    assert(target || dump_cb);
     assert(cb);
 
     if ((on_source_error == BLOCKDEV_ON_ERROR_STOP ||
@@ -382,10 +400,12 @@
         return;
     }
 
+    job->dump_cb = dump_cb;
     job->on_source_error = on_source_error;
     job->on_target_error = on_target_error;
     job->target = target;
     job->sync_mode = sync_mode;
+    job->common.paused = paused;
     job->common.len = len;
     job->common.co = qemu_coroutine_create(backup_run);
     qemu_coroutine_enter(job->common.co, job);
Index: new/blockdev.c
===================================================================
--- new.orig/blockdev.c	2014-07-16 11:09:01.000000000 +0200
+++ new/blockdev.c	2014-07-16 11:10:45.000000000 +0200
@@ -2115,7 +2115,7 @@
     }
 
     backup_start(bs, target_bs, speed, sync, on_source_error, on_target_error,
-                 block_job_cb, bs, &local_err);
+                 NULL, block_job_cb, bs, false, &local_err);
     if (local_err != NULL) {
         bdrv_unref(target_bs);
         error_propagate(errp, local_err);
Index: new/include/block/block_int.h
===================================================================
--- new.orig/include/block/block_int.h	2014-07-16 11:09:01.000000000 +0200
+++ new/include/block/block_int.h	2014-07-16 11:10:45.000000000 +0200
@@ -56,6 +56,9 @@
 #define BLOCK_OPT_REDUNDANCY        "redundancy"
 #define BLOCK_OPT_NOCOW             "nocow"
 
+typedef int BackupDumpFunc(void *opaque, BlockDriverState *bs,
+                           int64_t sector_num, int n_sectors, unsigned char *buf);
+
 typedef struct BdrvTrackedRequest {
     BlockDriverState *bs;
     int64_t offset;
@@ -539,7 +542,9 @@
                   int64_t speed, MirrorSyncMode sync_mode,
                   BlockdevOnError on_source_error,
                   BlockdevOnError on_target_error,
+                  BackupDumpFunc *dump_cb,
                   BlockDriverCompletionFunc *cb, void *opaque,
+                  bool paused,
                   Error **errp);
 
 #endif /* BLOCK_INT_H */
