Index: new/hw/net/e1000.c
===================================================================
--- new.orig/hw/net/e1000.c	2014-07-16 11:53:55.000000000 +0200
+++ new/hw/net/e1000.c	2014-07-16 12:01:38.000000000 +0200
@@ -1590,7 +1590,7 @@
 
     k->init = pci_e1000_init;
     k->exit = pci_e1000_uninit;
-    k->romfile = "efi-e1000.rom";
+    k->romfile = "pxe-e1000.rom";
     k->vendor_id = PCI_VENDOR_ID_INTEL;
     k->device_id = info->device_id;
     k->revision = info->revision;
Index: new/hw/net/ne2000.c
===================================================================
--- new.orig/hw/net/ne2000.c	2014-07-16 11:53:55.000000000 +0200
+++ new/hw/net/ne2000.c	2014-07-16 12:01:38.000000000 +0200
@@ -765,7 +765,7 @@
 
     k->init = pci_ne2000_init;
     k->exit = pci_ne2000_exit;
-    k->romfile = "efi-ne2k_pci.rom",
+    k->romfile = "pxe-ne2k_pci.rom",
     k->vendor_id = PCI_VENDOR_ID_REALTEK;
     k->device_id = PCI_DEVICE_ID_REALTEK_8029;
     k->class_id = PCI_CLASS_NETWORK_ETHERNET;
Index: new/hw/net/pcnet-pci.c
===================================================================
--- new.orig/hw/net/pcnet-pci.c	2014-07-16 11:53:55.000000000 +0200
+++ new/hw/net/pcnet-pci.c	2014-07-16 12:01:38.000000000 +0200
@@ -358,7 +358,7 @@
 
     k->init = pci_pcnet_init;
     k->exit = pci_pcnet_uninit;
-    k->romfile = "efi-pcnet.rom",
+    k->romfile = "pxe-pcnet.rom",
     k->vendor_id = PCI_VENDOR_ID_AMD;
     k->device_id = PCI_DEVICE_ID_AMD_LANCE;
     k->revision = 0x10;
Index: new/hw/net/rtl8139.c
===================================================================
--- new.orig/hw/net/rtl8139.c	2014-07-16 11:53:55.000000000 +0200
+++ new/hw/net/rtl8139.c	2014-07-16 12:01:38.000000000 +0200
@@ -3557,7 +3557,7 @@
 
     k->init = pci_rtl8139_init;
     k->exit = pci_rtl8139_uninit;
-    k->romfile = "efi-rtl8139.rom";
+    k->romfile = "pxe-rtl8139.rom";
     k->vendor_id = PCI_VENDOR_ID_REALTEK;
     k->device_id = PCI_DEVICE_ID_REALTEK_8139;
     k->revision = RTL8139_PCI_REVID; /* >=0x20 is for 8139C+ */
Index: new/hw/virtio/virtio-pci.c
===================================================================
--- new.orig/hw/virtio/virtio-pci.c	2014-07-16 11:53:55.000000000 +0200
+++ new/hw/virtio/virtio-pci.c	2014-07-16 12:01:38.000000000 +0200
@@ -1444,7 +1444,7 @@
     PCIDeviceClass *k = PCI_DEVICE_CLASS(klass);
     VirtioPCIClass *vpciklass = VIRTIO_PCI_CLASS(klass);
 
-    k->romfile = "efi-virtio.rom";
+    k->romfile = "pxe-virtio.rom";
     k->vendor_id = PCI_VENDOR_ID_REDHAT_QUMRANET;
     k->device_id = PCI_DEVICE_ID_VIRTIO_NET;
     k->revision = VIRTIO_PCI_ABI_VERSION;
