Index: new/vl.c
===================================================================
--- new.orig/vl.c	2014-06-17 06:14:54.000000000 +0200
+++ new/vl.c	2014-06-17 06:14:57.000000000 +0200
@@ -2754,8 +2754,8 @@
 
     p = qemu_opt_get(qemu_get_machine_opts(), "accel");
     if (p == NULL) {
-        /* Use the default "accelerator", tcg */
-        p = "tcg";
+        /* Use the default "accelerator", kvm */
+        p = "kvm";
     }
 
     while (!accel_initialised && *p != '\0') {
