Index: new/accel.c
===================================================================
--- new.orig/accel.c	2014-11-20 06:45:05.000000000 +0100
+++ new/accel.c	2014-11-20 07:25:45.000000000 +0100
@@ -87,8 +87,8 @@
 
     p = qemu_opt_get(qemu_get_machine_opts(), "accel");
     if (p == NULL) {
-        /* Use the default "accelerator", tcg */
-        p = "tcg";
+        /* Use the default "accelerator", kvm */
+        p = "kvm";
     }
 
     while (!accel_initialised && *p != '\0') {
