Index: new/qemu-img.c
===================================================================
--- new.orig/qemu-img.c	2013-11-27 12:00:36.000000000 +0100
+++ new/qemu-img.c	2013-11-27 12:06:45.000000000 +0100
@@ -1778,7 +1778,8 @@
 
     list = collect_image_info_list(filename, fmt, chain);
     if (!list) {
-        return 1;
+ 	// return success if snapshot does not exists
+        return 0;
     }
 
     switch (output_format) {
