Index: new/qemu-char.c
===================================================================
--- new.orig/qemu-char.c	2012-11-21 11:11:22.000000000 +0100
+++ new/qemu-char.c	2012-11-23 07:26:49.000000000 +0100
@@ -134,9 +134,9 @@
 void qemu_chr_generic_open(CharDriverState *s)
 {
     if (s->open_timer == NULL) {
-        s->open_timer = qemu_new_timer_ms(vm_clock,
+        s->open_timer = qemu_new_timer_ms(rt_clock,
                                           qemu_chr_fire_open_event, s);
-        qemu_mod_timer(s->open_timer, qemu_get_clock_ms(vm_clock) - 1);
+        qemu_mod_timer(s->open_timer, qemu_get_clock_ms(rt_clock) - 1);
     }
 }
 
