Index: new/qemu-img.c
===================================================================
--- new.orig/qemu-img.c	2014-05-05 06:55:26.000000000 +0200
+++ new/qemu-img.c	2014-05-05 07:33:18.000000000 +0200
@@ -1897,7 +1897,8 @@
 
     list = collect_image_info_list(filename, fmt, chain);
     if (!list) {
-        return 1;
+ 	// return success if snapshot does not exists
+        return 0;
     }
 
     switch (output_format) {
