Index: new/qemu-img.c
===================================================================
--- new.orig/qemu-img.c	2014-07-16 11:53:56.000000000 +0200
+++ new/qemu-img.c	2014-07-16 12:01:29.000000000 +0200
@@ -1931,7 +1931,8 @@
 
     list = collect_image_info_list(filename, fmt, chain);
     if (!list) {
-        return 1;
+ 	// return success if snapshot does not exists
+        return 0;
     }
 
     switch (output_format) {
