Index: new/qemu-img.c
===================================================================
--- new.orig/qemu-img.c	2014-06-17 06:14:28.000000000 +0200
+++ new/qemu-img.c	2014-06-17 06:14:56.000000000 +0200
@@ -1939,7 +1939,8 @@
 
     list = collect_image_info_list(filename, fmt, chain);
     if (!list) {
-        return 1;
+ 	// return success if snapshot does not exists
+        return 0;
     }
 
     switch (output_format) {
