Index: new/hw/i386/pc.c
===================================================================
--- new.orig/hw/i386/pc.c	2014-07-16 11:53:56.000000000 +0200
+++ new/hw/i386/pc.c	2014-07-16 12:01:30.000000000 +0200
@@ -1023,9 +1023,9 @@
     /* init CPUs */
     if (cpu_model == NULL) {
 #ifdef TARGET_X86_64
-        cpu_model = "qemu64";
+        cpu_model = "kvm64";
 #else
-        cpu_model = "qemu32";
+        cpu_model = "kvm32";
 #endif
     }
     current_cpu_model = cpu_model;
