Index: new/hw/i386/pc.c
===================================================================
--- new.orig/hw/i386/pc.c	2013-11-27 10:51:33.000000000 +0100
+++ new/hw/i386/pc.c	2013-11-27 11:37:06.000000000 +0100
@@ -984,9 +984,9 @@
     /* init CPUs */
     if (cpu_model == NULL) {
 #ifdef TARGET_X86_64
-        cpu_model = "qemu64";
+        cpu_model = "kvm64";
 #else
-        cpu_model = "qemu32";
+        cpu_model = "kvm32";
 #endif
     }
     current_cpu_model = cpu_model;
