Index: new/ui/spice-core.c
===================================================================
--- new.orig/ui/spice-core.c	2013-11-27 11:43:08.000000000 +0100
+++ new/ui/spice-core.c	2013-11-27 11:46:26.000000000 +0100
@@ -672,32 +672,35 @@
 
     if (tls_port) {
         x509_dir = qemu_opt_get(opts, "x509-dir");
-        if (NULL == x509_dir) {
-            x509_dir = ".";
-        }
 
         str = qemu_opt_get(opts, "x509-key-file");
         if (str) {
             x509_key_file = g_strdup(str);
-        } else {
+        } else if (x509_dir) {
             x509_key_file = g_strdup_printf("%s/%s", x509_dir,
                                             X509_SERVER_KEY_FILE);
+        } else {
+            x509_key_file = g_strdup("/etc/pve/local/pve-ssl.key");
         }
 
         str = qemu_opt_get(opts, "x509-cert-file");
         if (str) {
             x509_cert_file = g_strdup(str);
-        } else {
+        } else if (x509_dir) {
             x509_cert_file = g_strdup_printf("%s/%s", x509_dir,
                                              X509_SERVER_CERT_FILE);
+        } else {
+            x509_cert_file = g_strdup("/etc/pve/local/pve-ssl.pem");
         }
 
         str = qemu_opt_get(opts, "x509-cacert-file");
         if (str) {
             x509_cacert_file = g_strdup(str);
-        } else {
+        } else if (x509_dir) {
             x509_cacert_file = g_strdup_printf("%s/%s", x509_dir,
                                                X509_CA_CERT_FILE);
+        } else {
+            x509_cacert_file = g_strdup("/etc/pve/pve-root-ca.pem");
         }
 
         x509_key_password = qemu_opt_get(opts, "x509-key-password");
