From 861cf626eca5abb2fe7bc3be719acbf7086b94bb Mon Sep 17 00:00:00 2001
From: Derumier Alexandre <aderumier@odiso.com>
Date: Thu, 1 Mar 2012 17:00:30 +0100
Subject: [PATCH] update target to rhel 6.2 version


Signed-off-by: Derumier Alexandre <aderumier@odiso.com>
---
 sysconfigs/target/cpus-x86_64.conf |   36 ++++++++++++++++++++++++++++++---
 1 files changed, 32 insertions(+), 4 deletions(-)

Index: new/sysconfigs/target/cpus-x86_64.conf
===================================================================
--- new.orig/sysconfigs/target/cpus-x86_64.conf	2012-09-24 07:15:00.000000000 +0200
+++ new/sysconfigs/target/cpus-x86_64.conf	2012-09-24 07:15:25.000000000 +0200
@@ -1,11 +1,39 @@
 # x86 CPU MODELS
 
 [cpudef]
+   name = "cpu64-rhel6"
+   level = "4"
+   vendor = "AuthenticAMD"
+   family = "6"
+   model = "13"
+   stepping = "3"
+   feature_edx = "sse2 sse fxsr mmx clflush pse36 pat cmov mca pge mtrr sep apic cx8 mce pae msr tsc pse de fpu"
+   feature_ecx = "cx16 sse3 x2apic"
+   extfeature_edx = "lm fxsr mmx nx pat cmov pge syscall apic cx8 mce pae msr tsc pse de fpu"
+   extfeature_ecx = "sse4a abm svm lahf_lm"
+   xlevel = "0x8000000A"
+   model_id = "QEMU Virtual CPU version (cpu64-rhel6)"
+
+[cpudef]
+   name = "cpu64-rhel5"
+   level = "4"
+   vendor = "AuthenticAMD"
+   family = "6"
+   model = "6"
+   stepping = "3"
+   feature_edx = "sse2 sse fxsr mmx clflush pse36 pat cmov mca pge mtrr sep apic cx8 mce pae msr tsc pse de fpu"
+   feature_ecx = "sse3"
+   extfeature_edx = "3dnow 3dnowext lm fxsr mmx nx pat cmov pge syscall apic cx8 mce pae msr tsc pse de fpu"
+   extfeature_ecx = "svm"
+   xlevel = "0x8000000A"
+   model_id = "QEMU Virtual CPU version (cpu64-rhel5)"
+
+[cpudef]
    name = "Conroe"
    level = "2"
    vendor = "GenuineIntel"
    family = "6"
-   model = "2"
+   model = "15"
    stepping = "3"
    feature_edx = "sse2 sse fxsr mmx clflush pse36 pat cmov mca pge mtrr sep apic cx8 mce pae msr tsc pse de fpu"
    feature_ecx = "ssse3 sse3"
@@ -19,7 +47,7 @@
    level = "2"
    vendor = "GenuineIntel"
    family = "6"
-   model = "2"
+   model = "23"
    stepping = "3"
    feature_edx = "sse2 sse fxsr mmx clflush pse36 pat cmov mca pge mtrr sep apic cx8 mce pae msr tsc pse de fpu"
    feature_ecx = "sse4.1 cx16 ssse3 sse3"
@@ -33,7 +61,7 @@
    level = "2"
    vendor = "GenuineIntel"
    family = "6"
-   model = "2"
+   model = "26"
    stepping = "3"
    feature_edx = "sse2 sse fxsr mmx clflush pse36 pat cmov mca pge mtrr sep apic cx8 mce pae msr tsc pse de fpu"
    feature_ecx = "popcnt sse4.2 sse4.1 cx16 ssse3 sse3"
@@ -44,7 +72,7 @@
 
 [cpudef]
    name = "Westmere"
-   level = "11"
+   level = "2"
    vendor = "GenuineIntel"
    family = "6"
    model = "44"
