From f818a6aa247b71e8185a3277910ac1cd935432d1 Mon Sep 17 00:00:00 2001
From: Alexandre Derumier <aderumier@odiso.com>
Date: Mon, 9 Mar 2015 10:18:01 +0100
Subject: [PATCH] info max_mem : get_current_ram_size


Signed-off-by: Alexandre Derumier <aderumier@odiso.com>
---
 hw/virtio/virtio-balloon.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/hw/virtio/virtio-balloon.c b/hw/virtio/virtio-balloon.c
index df3333c..7f6da3f 100644
--- a/hw/virtio/virtio-balloon.c
+++ b/hw/virtio/virtio-balloon.c
@@ -316,7 +316,7 @@ static void virtio_balloon_stat(void *opaque, BalloonInfo *info)
     VirtIOBalloon *dev = opaque;
     info->actual = get_current_ram_size() - ((uint64_t) dev->actual <<
                                              VIRTIO_BALLOON_PFN_SHIFT);
-    info->max_mem = ram_size;
+    info->max_mem = get_current_ram_size();
 
     if (!(balloon_stats_enabled(dev) && balloon_stats_supported(dev) &&
 	    dev->stats_last_update)) {
-- 
1.7.10.4

