Index: kvm-86/configure
===================================================================
--- kvm-86.orig/configure	2009-05-22 09:47:55.000000000 +0200
+++ kvm-86/configure	2009-05-22 09:48:26.000000000 +0200
@@ -1393,8 +1393,8 @@
       prefix="/usr/local"
   fi
   mansuffix="/share/man"
-  datasuffix="/share/qemu"
-  docsuffix="/share/doc/qemu"
+  datasuffix="/share/kvm"
+  docsuffix="/share/doc/pve-qemu-kvm"
   binsuffix="/bin"
 fi
 
Index: kvm-86/net.h
===================================================================
--- kvm-86.orig/net.h	2009-05-22 09:47:48.000000000 +0200
+++ kvm-86/net.h	2009-05-22 09:48:26.000000000 +0200
@@ -124,8 +124,8 @@
 void net_host_device_add(Monitor *mon, const char *device, const char *opts);
 void net_host_device_remove(Monitor *mon, int vlan_id, const char *device);
 
-#define DEFAULT_NETWORK_SCRIPT "/etc/qemu-ifup"
-#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/qemu-ifdown"
+#define DEFAULT_NETWORK_SCRIPT "/etc/kvm/kvm-ifup"
+#define DEFAULT_NETWORK_DOWN_SCRIPT "/etc/kvm/kvm-ifdown"
 #ifdef __sun__
 #define SMBD_COMMAND "/usr/sfw/sbin/smbd"
 #else
